### PR TITLE
feat(postgres): Define application name for Postgres connection

### PR DESCRIPTION
**Pull request by rldhont**
## Description

The name of the application that connects to PostgreSQL is always QGIS even if it is QGIS Server or another application that uses the QGIS API.

We propose to be able to modify the name of the application that is used by QGIS to connect to PostgreSQL according to the configuration used.
This will make it easier to differentiate between QGIS Server and QGIS Desktop in PostgreSQL logs. It will also be possible to differentiate profiles, installations, or others in PostgreSQL logs.

The settings will be `application_name` in `PostgreSQL` section.

Funded by 3liz https://3liz.com

### CODE REVIEW

**Comment by rldhont** [comment]
The `test_core_networkaccessmanager` is in timeout.

File: src/providers/postgres/qgspostgresconn.cpp [hunk 1]
```diff
@@ -413,7 +414,9 @@ QgsPostgresConn::QgsPostgresConn( const QString &conninfo, bool readOnly, bool s
 
   if ( mPostgresqlVersion >= 90000 )
   {
-    LoggedPQexecNR( "QgsPostgresConn", QStringLiteral( "SET application_name='QGIS'" ) );
+    QgsSettings settings;
+    const QString applicationName = settings.value( QStringLiteral( "PostgreSQL/application_name" ), PG_APPLICATION_NAME, QgsSettings::Providers ).toString();
+    LoggedPQexecNR( "QgsPostgresConn", QStringLiteral( "SET application_name='%s'" ).arg( applicationName ) );
     LoggedPQexecNR( "QgsPostgresConn", QStringLiteral( "SET extra_float_digits=3" ) );
```
Comment (MorriganR) [inline src/providers/postgres/qgspostgresconn.cpp:420]: `LoggedPQexecNR( "QgsPostgresConn", QStringLiteral( "SET application_name='%s';SET extra_float_digits=3" ).arg( applicationName ) );`

Combining several SQL queries together will slightly reduce the negative impact of network delays on the speed of QGIS with the database.
Downside: The code becomes a bit harder to understand.

**Comment by rldhont** [comment]
Not staled, just waiting for reviews.

**Review by Gustry — APPROVED** [review]

File: src/providers/postgres/qgspostgresconn.cpp [hunk 2]
```diff
@@ -413,8 +414,9 @@ QgsPostgresConn::QgsPostgresConn( const QString &conninfo, bool readOnly, bool s
 
   if ( mPostgresqlVersion >= 90000 )
   {
-    LoggedPQexecNR( "QgsPostgresConn", QStringLiteral( "SET application_name='QGIS'" ) );
-    LoggedPQexecNR( "QgsPostgresConn", QStringLiteral( "SET extra_float_digits=3" ) );
+    QgsSettings settings;
+    const QString applicationName = settings.value( QStringLiteral( "PostgreSQL/application_name" ), PG_APPLICATION_NAME, QgsSettings::Providers ).toString();
+    LoggedPQexecNR( "QgsPostgresConn", QStringLiteral( "SET application_name=%s; SET extra_float_digits=3" ).arg( quotedValue( applicationName ) ) );
```
Comment (elpaso) [inline src/providers/postgres/qgspostgresconn.cpp:419]: `SET extra_float_digits=3`
is this related to settings the application name? 
What it does?

Comment (elpaso) [reply src/providers/postgres/qgspostgresconn.cpp:419]: Besides, shouldn't be better to default to  QCoreApplication::applicationName() plus QgsApplication::platform() ?

Comment (Gustry) [reply src/providers/postgres/qgspostgresconn.cpp:419]: The `SET extra_float_digits` was already there before this PR. It was a suggestion from @MorriganR to combine these 2 queries together https://github.com/qgis/QGIS/pull/49468#discussion_r929671570

Using `QgsApplication::platform()` might be a good idea for the default value indeed.
For the PR, they are several QGIS server running, we would like to be able to distinguish them.

Comment (rldhont) [reply src/providers/postgres/qgspostgresconn.cpp]: Hi @elpaso thanks for the review.

The SET extra_float_digits was already there before this PR. It was a suggestion from @MorriganR to combine these 2 queries together https://github.com/qgis/QGIS/pull/49468#discussion_r929671570

Using `QgsApplication::platform()` for default application name is a good idea. I have do it.

Comment (elpaso) [reply src/providers/postgres/qgspostgresconn.cpp]: You can also use  QCoreApplication::applicationName() to get "QGIS" (through QgsApplication::QGIS_APPLICATION_NAME) , there is no need for another static variable.

But I have another concern here:

if this is an option to be used in QGIS Desktop you probably need to expose it to the GUI.
if this is an option to be used in QGIS Server you need to expose it as a new QGIS_SERVER_.... env variable config option and document it.

Generally speaking, I understand your use case, I'm only doubtful about how widely used this option will be (especially if it was designed for the server context where our user base is rather narrow), I am just worried about adding another features to QGIS that nobody (except the authors) will use.

Note that I might be completely wrong here and this is one of the most wanted features.

**Comment by rldhont** [comment]
All tests are green except MingW like in master

File: src/providers/postgres/qgspostgresconn.cpp (hunk 2 above)
Comment (rldhont) [reply src/providers/postgres/qgspostgresconn.cpp]: I have check `QCoreApplication::applicationName() ` and it returns `QGIS3` not only `QGIS`. https://github.com/qgis/QGIS/blob/0c577f07ba45e3903eb09b5239d2d5da572e8f6c/src/core/qgsapplication.cpp#L152

Then I have made `application_name` an option in QGIS3.conf file like the PostgreSQL connection time_out. You don't have any GUI or any env varaible for connection timeout config. I was thinking that the application_name is much more something to configure at the install. For me it depends on organization and way to install QGIS or QGIS Server.

@elpaso do you know where the PostgreSQL provider can be configure ?

Comment (elpaso) [reply src/providers/postgres/qgspostgresconn.cpp]: > I have check `QCoreApplication::applicationName() ` and it returns `QGIS3` not only `QGIS`.

Is that a problem?

> 
> https://github.com/qgis/QGIS/blob/0c577f07ba45e3903eb09b5239d2d5da572e8f6c/src/core/qgsapplication.cpp#L152
> 
> Then I have made `application_name` an option in QGIS3.conf file like the PostgreSQL connection time_out. You don't have any GUI or any env varaible for connection timeout config. I was thinking that the application_name is much more something to configure at the install. For me it depends on organization and way to install QGIS or QGIS Server.

In general we should follow the 12 factors paradigm (https://12factor.net/) which mandates config through the environment.

We should definitely think of a generic way to override QGIS3.ini settings through the env if we want to become serious about web map services deployment on complex infrastructures.

> 
> @elpaso do you know where the PostgreSQL provider can be configure ?

There isn't any except per-connection settings. IMHO this is a legacy of the server being created on top of the QGIS desktop application.

Comment (elpaso) [reply src/providers/postgres/qgspostgresconn.cpp]: > The `SET extra_float_digits` was already there before this PR. It was a suggestion from @MorriganR to combine these 2 queries together [#49468 (comment)](https://github.com/qgis/QGIS/pull/49468#discussion_r929671570)
> 

Fair enough, a comment in the code wouldn't hurt.

**Comment by rldhont** [comment]
Hi @elpaso, I have the requested changes.

About the 12 factors paradigm and the way to override QGIS3.ini settings through the env, do you have some recommendation about variable name ?

**Comment by elpaso** [comment]
> Hi @elpaso, I have the requested changes.
> 
> About the 12 factors paradigm and the way to override QGIS3.ini settings through the env, do you have some recommendation about variable name ?

`QGIS_SERVER_APPLICATION_NAME` ?

**Comment by rldhont** [comment]
@elpaso I have added the variable QGIS_SERVER_APPLICATION_NAME

**Comment by rldhont** [comment]
The PR is not stale. I have rebased the branch with master and waiting for a last review.

@elpaso is it good for you ?

File: src/server/qgsserversettings.cpp [hunk 3]
```diff
@@ -379,6 +379,16 @@ void QgsServerSettings::initSettings()
                                          };
   mSettings[ sAllowedExtraSqlTokens.envVar ] = sAllowedExtraSqlTokens;
 
+  const Setting sApplicationName = { QgsServerSettingsEnv::QGIS_SERVER_APPLICATION_NAME,
+                                     QgsServerSettingsEnv::DEFAULT_VALUE,
+                                     QStringLiteral( "The QGIS Server application name" ),
+                                     QStringLiteral( "/qgis/server_application_name" ),
```
Comment (elpaso) [inline src/server/qgsserversettings.cpp:385]: Shouldn't this be the same used in the settings ("PostgreSQL/application_name")?

**Comment by elpaso** [comment]
> The PR is not stale. I have rebased the branch with master and waiting for a last review.
> 
> @elpaso is it good for you ?

Just one thing: this setting specific to postgres but is there any change that it will be used more widely?

(I assumed the second, that's why I suggested `QGIS_SERVER_APPLICATION_NAME`)

If the usage is (and presumably will ever be)  postgres-only I'd probably recommend to add `POSTGRES` (or `PG` or similar) to the env var name, in any case, make sure that there is consistency in the naming of the env variable and the settings entry (mention postgres in all or them or none).

**Comment by elpaso** [comment]
.... a possible maybe silly use case: when using multiple instances of the server you might want to add the application name in the logs to differentiate them?
